### ISSUE TITLE
Terrible error message when trying to use borrowed `self` in `move` closure

### ISSUE THREAD

**Issue by cramertj** [rust-lang/rust#45634]
The code:
```rust
struct Foo {
    name: String,
}

impl Foo {
    fn foo(&self) -> Box<Fn()> {
        Box::new(move || println!("{}", self.name))
    }
}
```

Expected behavior: prints error that looks something like this:
```rust
self.name
^^^^ cannot use self in 'static move closure because `self` is borrowed
```

Actual behavior: prints this error message:
```rust
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> src/main.rs:8:9
  |
8 |         Box::new(move || println!("{}", self.name))
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 7:5...
 --> src/main.rs:7:5
  |
7 | /     fn foo(&self) -> Box<Fn()> {
8 | |         Box::new(move || println!("{}", self.name))
9 | |     }
  | |_____^
note: ...so that the type `[closure@src/main.rs:8:18: 8:51 self:&Foo]` will meet its required lifetime bounds
 --> src/main.rs:8:9
  |
8 |         Box::new(move || println!("{}", self.name))
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: but, the lifetime must be valid for the static lifetime...
note: ...so that expression is assignable (expected std::boxed::Box<std::ops::Fn() + 'static>, found std::boxed::Box<std::ops::Fn()>)
 --> src/main.rs:8:9
  |
8 |         Box::new(move || println!("{}", self.name))
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by cramertj** [comment]
This issue isn't limited to `self`, either, but this was the case in which the error originally occurred. While I understand it might be difficult to identify exactly which variable in the closure has too short of a lifetime, it should be possible to at least give a better error message like "value with non-`'static` lifetime referenced in static closure".

**Comment by estebank** [comment]
Triage: very little change:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> src/main.rs:7:9
  |
7 |         Box::new(move || println!("{}", self.name))
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 6:5...
 --> src/main.rs:6:5
  |
6 | /     fn foo(&self) -> Box<Fn()> {
7 | |         Box::new(move || println!("{}", self.name))
8 | |     }
  | |_____^
note: ...so that the type `[closure@src/main.rs:7:18: 7:51 self:&Foo]` will meet its required lifetime bounds
 --> src/main.rs:7:9
  |
7 |         Box::new(move || println!("{}", self.name))
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: but, the lifetime must be valid for the static lifetime...
  = note: ...so that the expression is assignable:
          expected std::boxed::Box<std::ops::Fn() + 'static>
             found std::boxed::Box<std::ops::Fn()>
```

Error should point at `self` inside the closure, and to the closure.

**Comment by estebank** [comment]
Ast output:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> file.rs:7:18
  |
7 |         Box::new(move || println!("{}", self.name))
  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 6:5...
 --> file.rs:6:5
  |
6 | /     fn foo(&self) -> Box<Fn()> {
7 | |         Box::new(move || println!("{}", self.name))
8 | |     }
  | |_____^
  = note: ...so that the types are compatible:
          expected &Foo
             found &Foo
  = note: but, the lifetime must be valid for the static lifetime...
  = note: ...so that the expression is assignable:
          expected std::boxed::Box<(dyn std::ops::Fn() + 'static)>
             found std::boxed::Box<dyn std::ops::Fn()>
```

MIR output:
```
error: lifetime may not live long enough
 --> file.rs:7:9
  |
6 |     fn foo(&self) -> Box<Fn()> {
  |            - let's call the lifetime of this reference `'1`
7 |         Box::new(move || println!("{}", self.name))
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'static`
```

**Comment by estebank** [comment]
No change in the output. The compiler should suggest a signature of `fn foo(&self) -> Box<dyn Fn() + '_>` when we infer a `'static` lifetime, as it is rarely what we desire. We already provide that suggestion for `impl Trait`s in the return type.

**Comment by cramertj** [comment]
@estebank FWIW in the case this was originally minimized from, the solution was actually to call `self.name.clone()` and then move the resulting name into the closure, rather than moving `self` in . I don't think a suggestion is what's needed here-- I think the primary thing that would help would be a hint that the use of `self.name` in the closure is what's causing the lifetime to be restricted to `'1`.

**Comment by estebank** [comment]
Closing in favor of the newer ticket, which will be closed by #106641.
